### ISSUE TITLE
[Bug]: const ref string local variables will sometimes leak memory when used in ternary expressions

### ISSUE THREAD

**Issue by lydia-duncan** [chapel-lang/chapel#25417]
### Summary of Problem

**Description:**
Discovered when testing the solution for #25032 (which improved the situation when such variables were returned and when the string result of a ternary was declared as var and not returned).

I can't fully enumerate why each case does or does not leak memory, but it does seem to at least require that the string variable is declared `const ref` (or probably also `ref`) and that it is involved in an if expression in some way.  I was not seeing the same behavior for records, or even records that contained strings.

It's definitely related to callDestructors, see the PR message in #25384 

We think this will be significantly easier to fix and avoid when callDestructors is migrated as part of the dyno effort.

**Is this issue currently blocking your progress?**
not mine, I was investigating a related bug

### Steps to Reproduce

**Source Code:**

E.g.

```chapel
proc positive(z : string) {
  const ref em = '';  // Fine (possibly because it is empty)
  const ref sq = '~'; // This is associated with a leak
  const ref sp = ' ';  // This is associated with a leak
  const ref pl = '+';  // This is associated with a leak
  const ref vr = z;    // Fine because the memory is cleaned up when the returned value is cleaned up
  return if vr == sq then sp else (if vr == pl then vr else em);
}

const s = positive('+');
writeln(s);

```

**Compile command:**
`chpl foo.chpl`

**Execution command:**
`./foo --memLeaks`

**Associated Future Test(s):**
[`test/types/string/ternaryReturn4.chpl`](https://github.com/chapel-lang/chapel/blob/main/test/types/string/ternaryReturn4.chpl) #25384
[`test/types/string/ternaryPlainRef.chpl`](https://github.com/chapel-lang/chapel/blob/main/test/types/string/ternaryPlainRef.chpl) #25384

### Configuration Information

- Output of `chpl --version`: chapel 2.2 (pre-release)
- Output of `$CHPL_HOME/util/printchplenv --anonymize`:
CHPL_TARGET_PLATFORM: darwin
CHPL_TARGET_COMPILER: clang
CHPL_TARGET_ARCH: x86_64
CHPL_TARGET_CPU: native
CHPL_LOCALE_MODEL: flat
CHPL_COMM: none
CHPL_TASKS: qthreads
CHPL_LAUNCHER: none
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
CHPL_GMP: none
CHPL_HWLOC: bundled
CHPL_RE2: bundled
CHPL_LLVM: none *
CHPL_AUX_FILESYS: none
- Back-end compiler and version, e.g. `gcc --version` or `clang --version`:
Apple clang version 15.0.0 (clang-1500.1.0.2.5)
Target: x86_64-apple-darwin22.6.0
Thread model: posix
- (For Cray systems only) Output of `module list`: N/A

**Comment by lydia-duncan** [comment]
@mppf - would you like this added to dyno so we don't forget about it when dyno gets to callDestructors?  Or would you prefer it not get labeled that way, since it's a problem with the production compiler rather than dyno itself?

**Comment by e-kayrakli** [comment]
`const ref s = "some string";` doesn't feel right to me. What's a reference to a literal?

Trying it out, I see that you can also do

```chpl
const ref x = 5;
```

but you can't

```chpl
param x = 5;
const ref xRef = x;
```

as the latter gives

```
error: Cannot set a reference to a param variable.
```

If I can't set reference to a param, I don't think I should be able to do the same for a literal.

**Comment by mppf** [comment]
@e-kayrakli -- it's my understanding that the failure of `param x = 5; const ref xRef = x;` is a bug. I based that on the direction from issue #17195 implemented in PR #23055. (If we decided to allow literals passed to `const ref`, that should apply also to `param` values, and it should apply also to creating local `const ref`s outside of a call).

**Comment by e-kayrakli** [comment]
Hmm.. I was vaguely remembering that discussion thanks for pointing it out. Reading it a bit, I can't see the rationale for extending support for `const ref`s into variable declarations using literals as opposed to passing literals/params into formals. I would consider them to be separate things, but I also see this as a digression from the topic at hand here.

**Comment by bradcray** [comment]
I agree that permitting them to be passed to refs doesn't necessarily mean supporting them in declaration contexts as well—at least, I'm not sure it ever occurred to me that we were supporting both in the argument-passing conversation.  But I don't necessarily object to supporting those patterns either (though they do seem slightly weird).  I mostly agree that we should be consistent about the declaration forms in any case (i.e., permit all or disallow all).
